lib/charms/grafana_agent/v0/cos_machine.py

```diff
@@ -0,0 +1,230 @@
+            "dashboards": {
+                "dashboards": self._dashboards,
+            },
+        }
+
+        return json.dumps(data)
+
+    def _scrape_jobs(self) -> List[Dict]:
+        """Return a prometheus_scrape-like data structure for jobs."""
+        job_name_prefix = self._charm.app.name
+        return [
+            {"job_name": f"{job_name_prefix}_{key}", **endpoint}
+            for key, endpoint in enumerate(self._metrics_endpoints)
+        ]
+
+    def _metrics_alert_rules(self) -> Dict:
+        """Use (for now) the prometheus_scrape AlertRules to initialize this."""
+        alert_rules = MetricsAlerts()
+        alert_rules.add_path(self.metrics_rules_dir, recursive=self._recursive)
+        return alert_rules.as_dict()
+
+    def _log_alert_rules(self) -> Dict:
+        """Use (for now) the loki_push_api AlertRules to initialize this."""
+        alert_rules = LogAlerts()
+        alert_rules.add_path(self.log_rules_dir, recursive=self._recursive)
+        return alert_rules.as_dict()
+
+    def _dashboards(self) -> List[str]:
+        dashboards = []
+        for d in self._dashboard_dirs:
+            for path in Path(d).glob("*"):
+                dashboards.append(self._encode_dashboard_content(path.read_bytes()))
+
+        return dashboards
+
+    def _encode_dashboard_content(self, content: Union[str, bytes]) -> str:
+        if isinstance(content, str):
+            content = bytes(content, "utf-8")
+
+        return base64.b64encode(lzma.compress(content)).decode("utf-8")
+
+
+class CosMachineDataChanged(EventBase):
+    """Event emitted when a `CosMachineProvider` joins or updates data."""
+
+
+class CosMachineConsumerEvents(ObjectEvents):
+    """Event descriptor for events raised by `CosMachineConsumer`."""
+
+    data_changed = EventSource(CosMachineDataChanged)
+
+
+class CosMachineConsumer(Object):
+    """Provider class for the cos_machine interface."""
+
+    on = CosMachineConsumerEvents()
+
+    def __init__(
+        self,
+        charm,
+        relation_name: str = DEFAULT_RELATION_NAME,
+        refresh_events: Optional[List] = None,
+    ):
+        """Create a CosMachineConsumer instance.
+
+        Args:
+            charm: The `CharmBase` instance that is instantiating this object.
+            relation_name: The name of the relation to communicate over.
+            refresh_events: List of events on which to resfresh relation data.
+        """
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._relation_name = relation_name
+        self._refresh_events = refresh_events or [self._charm.on.config_changed]
+
+        events = self._charm.on[relation_name]
+        self.framework.observe(events.relation_joined, self._on_relation_data_changed)
+        self.framework.observe(events.relation_changed, self._on_relation_data_changed)
+        for event in self._refresh_events:
+            self.framework.observe(event, self.trigger_refresh)
```

Comment:
I'm not sure I understand why we would have `refresh_events` here. We don't write anything to relation data so what is there to refresh?

Reply:
I'm not 100% convinced about the `refresh_events` pattern... I would prefer a public method in the lib that can be called from the charm when it is needed.

Reply:
We don't have anything in relation data, but the events fired off do tell grafana-agent itself to update the relation data on the other side (propagating alerts, re-writing the config). 

The `refresh_events` pattern is useful/needed in k8s, because a unit may be killed/recycled and have its address change. It's not necessarily as useful here, because we're a subordinate and not looking "outside", but it's also almost impossible to know whether anything else has changed. 

Arguably, `on_start` is a better "refresh event" than `config-changed`, but we learned a long time ago that refreshing/restarting a charm may not send any meaningful events (relation data is the same, so no changes there) which will cause a rebuild of appropriate relation data.

Right now, we are relying on [observing `config-changed`](https://github.com/canonical/grafana-agent-k8s-operator/blob/machine_charm/src/grafana_agent.py#L115) to ensure that the config file is in a sane state, but this may not be reliable. The refresh event here is a safeguard, and while we could have a public method which can be called from the charm, the same essential question of "on which lifecycle events should we force a re-evaluation of what's already in relation data" remains. 

If the charm is restarted (machine killed, host dies, whatever), no alert events will fire (the relation data is the same), and [`upgrade_charm`](https://github.com/canonical/grafana-agent-k8s-operator/blob/machine_charm/src/grafana_agent.py#L99) will not fire (it's not an upgrade), so how are we sure we'll actually copy the data into the right place? It's this method, or performing the same logic from the subordinate itself, when it's needed (whenever that is, which will be the same event as here).

We can move it there, but we'll need the same flow in either case.